### PR TITLE
[pkg/queue] Add `StartConsumersWithFactory` function

### LINKED ISSUE #2685: [pkg/queue] Provide factory for consumers
## Requirement - what kind of business use case are you trying to solve?

As stated in open-telemetry/opentelemetry-collector#2254 for performance reasons we want to be able to keep state for each consumer of a bounded queue so that we can reuse a certain struct.

## Problem - what in Jaeger blocks you from solving the requirement?

It's not currently possible to define a state for each consumer of a bounded queue: the public interface for the bounded queue only allows for providing a function and every consumer runs that function, so you can only keep a global state for all consumers.

https://github.com/jaegertracing/jaeger/blob/1b209472dc6a35f7de90c7a9048cb4e960cc2335/pkg/queue/bounded_queue.go#L61

## Proposal - what do you suggest to solve the problem or improve the existing situation?

A possible solution would work like this.

1. Define an interface like
```golang
type Consumer interface {
  Consume(item interface{})
}
```

2. Provide a new way to start a bounded queue with signature
```golang
func (q *BoundedQueue) StartConsumersWithFactory(num int, factory func() Consumer)
```
Reimplement consumer logic internally so that each consumer is built using the factory.

3. Rewrite `StartConsumers` using this factory
```golang
type statelessConsumer struct{
  consumefn func(item interface{}) 
}

func (c *statelessConsumer) Consume(item interface{}){
  c.consumefn(item)
}

func (q *BoundedQueue) StartConsumers(num int, consumer func(item interface{})) { 
  consumer := &statelessConsumer{consumer}
  q.StartConsumersWithFactory(num, func(){
    return consumer 
  })
}
```

## Any open questions to address

I think this would keep backwards-compatible behavior with minimal/no impact to performance but I am not a Go expert so that should be double-checked.

### PR DESCRIPTION
**Pull request by mx-psi**
## Which problem is this PR solving?
- Fixes #2685

## Short description of the changes

- Add `StartConsumersWithFactory` function to allow for keeping state per-consumer.
- Rewrite `StartConsumers` with a dummy factory function.
- Add unit test and benchmark for the new function.

### CODE REVIEW

File: pkg/queue/bounded_queue_test.go [hunk 1]
```diff
@@ -112,6 +112,80 @@ func TestBoundedQueue(t *testing.T) {
 	assert.False(t, q.Produce("x"), "cannot push to closed queue")
 }
 
+func TestBoundedQueueWithFactory(t *testing.T) {
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue_test.go:115]: is this test essentially a duplicate of the previous one? It's a lot of duplication and complexity. I would rather refactor it into a helper function that can be used with both factory-based and callback-based initialization.

File: pkg/queue/bounded_queue.go [hunk 2]
```diff
@@ -89,6 +95,25 @@ func (q *BoundedQueue) StartConsumers(num int, consumer func(item interface{}))
...
 
+// statelessConsumer wraps a consume function callback
+type statelessConsumer struct {
+	consumefn func(item interface{})
+}
+
+// Consumer consumes an item from a bounded queue
+func (c *statelessConsumer) Consume(item interface{}) {
+	c.consumefn(item)
+}
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue.go:106]: I would do this more in line with stdlib/http/HandlerFunc
```suggestion
type ConsumerFunc func(item interface{})

// Consumer consumes an item from a bounded queue
func (c *ConsumerFunc) Consume(item interface{}) {
	c(item)
}
```

File: pkg/queue/bounded_queue_test.go [hunk 3]
```diff
@@ -30,33 +30,14 @@ import (
 	uatomic "go.uber.org/atomic"
 )
 
-// In this test we run a queue with capacity 1 and a single consumer.
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue_test.go:33]: I would keep the comment at the helper function, it describes more the helper than the actual tests.

File: pkg/queue/bounded_queue_test.go [hunk 4]
```diff
@@ -112,6 +93,58 @@ func TestBoundedQueue(t *testing.T) {
...
+		counter.Inc(1)
+	})
+	assert.Equal(t, 1, q.Capacity())
+
+	var startLock sync.Mutex
+
+	startLock.Lock() // block consumers
+	consumerState := newConsumerState(t)
+
+	q.StartConsumers(1, func(item interface{}) {
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue_test.go:113]: There's still a bunch of boiler plate across these tests. What if the helper function looked like this instead:
```go
func helper(t *testing.T, startConsumers func(q *BounderQueue, consumerFn func(item interface{})))
```

Then the tests become
```go
func TestBoundedQueue(t *testing.T) {
    helper(t, func (q *BounderQueue, consumerFn func(item interface{})) {
        q.StartConsumers(1, consumerFn)
    })
}

func TestBoundedQueueWithFactory(t *testing.T) {
    helper(t, func (q *BounderQueue, consumerFn func(item interface{})) {
        q.StartConsumersWithFactory(1, func () Consumer { return ConsumerFunc(consumerFn) })
    })
}
```

File: pkg/queue/bounded_queue.go (hunk 2 above)
Comment (mx-psi) [reply pkg/queue/bounded_queue.go:106]: That makes a lot of sense. Should this be exported?

Comment (mx-psi) [reply pkg/queue/bounded_queue.go]: Should be done, marking as solved

**Comment by mx-psi** [comment]
@yurishkuro Thanks for the comments, I think I addressed everything, hopefully this looks better :)

File: pkg/queue/bounded_queue_test.go [hunk 5]
```diff
@@ -33,7 +33,7 @@ import (
 // In this test we run a queue with capacity 1 and a single consumer.
 // We want to test the overflow behavior, so we block the consumer
 // by holding a startLock before submitting items to the queue.
-func TestBoundedQueue(t *testing.T) {
+func helper(t *testing.T, startConsumers func(q *BoundedQueue, consumerFn func(item interface{}))) {
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue_test.go:36]: hehe, I didn't mean the name `helper` literally, was just too lazy to type what you had, but that's fine.

File: pkg/queue/bounded_queue_test.go [hunk 6]
```diff
@@ -112,6 +112,18 @@ func TestBoundedQueue(t *testing.T) {
 	assert.False(t, q.Produce("x"), "cannot push to closed queue")
 }
 
+func TestBoundedQueue(t *testing.T) {
```
Comment (yurishkuro) [inline pkg/queue/bounded_queue_test.go:115]: much cleaner, minimal changes

**Review by yurishkuro — APPROVED** [review]
Thanks!

File: pkg/queue/bounded_queue_test.go (hunk 5 above)
Comment (mx-psi) [reply pkg/queue/bounded_queue_test.go:36]: hahaha, I couldn't come up with anything better so I left it like you said 😄

**Pull request merged**